0125-bootspec: initrd is optional, add system, extensions, clarify the meaning of optional (#7)
grahamc/rfcs: rfcs/0125-bootspec.md

{
  // Toplevel key describing the version of the specification used in the document
  "v1": {

>     // System for which the bootspec has been built, useful for architecture-dependent removable device fallback path

Reviewer: ```suggestion
    // System type the bootspec is intended for (e.g. `x86_64-linux`, `aarch64-linux`)
```

>     "system": "x86_64-linux",

Reviewer: I think we should also add `(Required) ...` to all the other fields, to make it extremely obvious which fields are optional, and which are required for a compliant implementation.

    // Path to the stage-2 init, executed by the initrd
    "init": "/nix/store/xxx-nixos-system-xxx/init",

>     // Optional path to the initrd

Reviewer: ```suggestion
    // (Optional) Path to the initrd
```

{
  // Toplevel key describing the version of the specification used in the document
  "v1": {
    // (Required) System type the bootspec is intended for (e.g. `x86_64-linux`, `aarch64-linux`)
    "system": "x86_64-linux",

    // (Required) Path to the stage-2 init, executed by the initrd
    "init": "/nix/store/xxx-nixos-system-xxx/init",

    // (Optional) Path to the initrd
    "initrd": "/nix/store/xxx-initrd-linux/initrd",

>     // (Optional) path to a tool to dynamically add secrets to an initrd

Reviewer: Nit:

```suggestion
    // (Optional) Path to a tool to dynamically add secrets to an initrd
```

      "loglevel=4"
    ],

    // (Required) The label of the system. It should contain the operating system, kernel version,
    // and other user-relevant information to identify the system. This corresponds
    // loosely to `config.system.nixos.label`.
    "label": "NixOS 21.11.20210810.dirty (Linux 5.15.30)",

    // (Required) Top level path of the closure, in case some spelunking is required
    "toplevel": "/nix/store/xxx-nixos-system-xxx",

    // (Optional) Mapping of specialisation names to their bootspec document
    "specialisation": {
      // <name> corresponds to <name> in specialisation.<name>.configuration.
      // Note that it is not valid for a specialisation to be a different version than the main document.
      "<name>": {
        // A full Bootspec v1 document, _without_ the `"v1"` key.
        // Note that it is not valid for a specialisation to have further specialisations.
      }
    },

    // (Optional) User-specified metadata
    "extensions": {
      // <namespace> corresponds to a user-provided namespace
      // to reduce the chances of collision when using

>       // multiple extensions at once

Reviewer: Nit: The reflowing looks a little weird here, when compared to the line lengths of the other comments (e.g. compared to https://github.com/grahamc/rfcs/pull/7/files#diff-709266622ec29f9cabb5a5b4e6f93792851db9cfebd72bdf78f3032b437d8764L148).

      "<namespace>": {
        // Any kind of representable JSON is valid here
      }
    }
  }
}
```

> An *optional* field mean: either missing or present, **not null**.

Reviewer: ```suggestion
An *optional* field means: a field that is either missing or present, but **never `null`**.
```

    // Top level path of the closure, in case some spelunking is required
    "toplevel": "/nix/store/xxx-nixos-system-xxx",

>     // Specialisations are optional

Reviewer: ```suggestion
    // (Optional) Mapping of specialisation names to their bootspec document
```

    "specialisation": {
      // <name> corresponds to <name> in specialisation.<name>.configuration.
      // Note that it is not valid for a specialisation to be a different version than the main document.
      "<name>": {
        // A full Bootspec v1 document, _without_ the `"v1"` key.
        // Note that it is not valid for a specialisation to have further specialisations.
      }
    },

>     // Optional user-specific metadata

Reviewer: ```suggestion
    // (Optional) User-specified metadata
```

    "extensions": {
      // <namespace> corresponds to a user-provided namespace
      // to reduce the chances of collision when using
      // multiple extensions at once
      "<namespace>": {
        // Any kind of representable JSON is valid here
      }
    }
  }
}
```

> An *optional* field mean: either missing or present, **not null**.

Reviewer: IMO, optional fields disallowing `null` is a bad idea. I don't think separating a missing field from a `null`-valued one is a good idea _for the spec_.

This is not talking about the custom data underneath `extensions` -- if consumers want to decide if `{"pce":null}` is different from `{}`, that's up to them; but the spec itself should _not_, IMO, care about `null` vs. unset.

`"init"` (right now) is required -- neither being unset, nor being `null` are valid. And so I think that, since `"extensions"` is not required, being unset _AND_ being `null` should be valid.

Author: We should go back then to the RFC-0125 PR discussion because this is something we decided during a meeting to disambiguate what was meant by "optional".

I agree with you that we should not spec what happens in the custom data, this is not the role of the spec and consumers can do whatever they want in the `HashMap<String, serde_json::Value>` implementation at the moment.

I don't necessarily agree that `extensions: null` should be a valid representation of `extensions: HashMap::new()`.

In JavaScript, `null` and `undefined` are two different values for `!==` (though they are equal for `==`), the idea behind constraining the values is to make it easier to reason in other languages such as those.

Making so that `null` never happen in the document is a nice way to eliminate this class of bugs IMHO.

Reviewer: > Making so that `null` never happen in the document is a nice way to eliminate this class of bugs IMHO.

You've convinced me with that single line; I think it's fine to deny `null`. :smile:
